### PR TITLE
build: Decrease `make codegen` time by (at least) 4 min

### PR DESCRIPTION
**Pull request by simster7**
Closes https://github.com/argoproj/argo-workflows/issues/5279

Before:
```
real	9m42.742s
user	13m24.102s
sys	10m12.749s
```

After:
```
real	5m36.731s
user	9m11.539s
sys	7m38.640s
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -95,14 +95,14 @@ PROTO_BINARIES := $(GOPATH)/bin/protoc-gen-gogo $(GOPATH)/bin/protoc-gen-gogofas
 
 # go_install,path
 define go_install
-	go mod vendor
+	[ -e ./vendor ] || go mod vendor
```
Comment (simster7) [inline Makefile:98]: Stops from running `go mod vendor` on every invocation

File: Makefile [hunk 2]
```diff
@@ -364,20 +364,19 @@ $(GOPATH)/bin/golangci-lint:
...
 	go mod tidy
 	# Lint Go files
 	$(GOPATH)/bin/golangci-lint run --fix --verbose --concurrency 4 --timeout 5m
-	# Lint UI files
-ifeq ($(STATIC_FILES),true)
-	yarn --cwd ui lint
-endif
+	# Lint swagger files
+	swagger validate api/openapi-spec/swagger.json
+	go test ./api/openapi-spec
```
Comment (simster7) [inline Makefile:375]: No need to validate on codegen, moved to `make lint`

File: Makefile [hunk 3]
```diff
@@ -512,30 +511,31 @@ clean:
...
+	    workflow:v1alpha1 \
+	    --go-header-file ./hack/custom-boilerplate.go.txt
+	[ -e ./v3 ] && rm -rf v3
 
 dist/kubernetes.swagger.json:
 	@mkdir -p dist
 	./hack/recurl.sh dist/kubernetes.swagger.json https://raw.githubusercontent.com/kubernetes/kubernetes/v1.17.5/api/openapi-spec/swagger.json
 
-pkg/apiclient/_.secondary.swagger.json: hack/swagger/secondaryswaggergen.go server/static/files.go pkg/apis/workflow/v1alpha1/openapi_generated.go dist/kubernetes.swagger.json
+pkg/apiclient/_.secondary.swagger.json: hack/swagger/secondaryswaggergen.go pkg/apis/workflow/v1alpha1/openapi_generated.go dist/kubernetes.swagger.json
```
Comment (simster7) [inline Makefile:538]: Static files not needed for codegen

**Comment by simster7** [comment]
Looking for more improvements

**Comment by simster7** [comment]
Working on failures...

File: Makefile [hunk 4]
```diff
@@ -571,8 +579,12 @@ docs/fields.md: api/openapi-spec/swagger.json $(shell find examples -type f) hac
 	env ARGO_SECURE=false ARGO_INSECURE_SKIP_VERIFY=false ARGO_SERVER= ARGO_INSTANCEID= go run ./hack docgen
 
 # generates several other files
-docs/cli/argo.md: $(CLI_PKGS) server/static/files.go hack/cli/main.go
+docs/cli/argo.md: $(CLI_PKGS) hack/cli/main.go
+    # If the static file does not exist, generate a mock file with required function signatures
+	[ -e ./server/static/files.go ] || echo -e "//DELETEME\npackage static\nimport \"net/http\"\nfunc ServeHTTP(w http.ResponseWriter, r *http.Request) {}\nfunc Hash(file string) string { return \"\" }" > ./server/static/files.go
```
Comment (simster7) [inline Makefile:584]: Nifty trick to save the need to spend 4 min generating static files JUST for these two function signatures

File: Makefile [hunk 5]
```diff
@@ -236,8 +236,13 @@ dist/argoexec.image: $(ARGOEXEC_PKGS)
 # generation
 
 .PHONY: codegen
-codegen: \
-	pkg/apis/workflow/v1alpha1/generated.proto \
+codegen: codegen-types codegen-swagger codegen-docs
```
Comment (simster7) [inline Makefile:239]: Individual targets

File: Makefile (hunk 4 above)
Comment (alexec) [reply Makefile:594]: can we use this everywhere?

Comment (alexec) [reply Makefile:594]: maybe create a stub file we can use?

./server/static/files.go.stub

Comment (simster7) [reply Makefile:594]: Yup, will rework

File: Makefile (hunk 5 above)
Comment (alexec) [reply Makefile:239]: could we have `make types` `make swagger` and `make docs`?

Comment (simster7) [reply Makefile:239]: Yup!

File: Makefile [hunk 6]
```diff
@@ -131,7 +131,7 @@ define docker_pull
 endef
 
 define gen-static-stub
-	[ -e ./server/static/files.go ] || echo -e "//DELETEME\npackage static\nimport \"net/http\"\nfunc ServeHTTP(w http.ResponseWriter, r *http.Request) {}\nfunc Hash(file string) string { return \"\" }" > ./server/static/files.go
+	[ -e ./server/static/files.go ] || cp ./server/static/files.go.stub ./server/static/files.go
```
Comment (alexec) [inline Makefile:134]: you can use `ifneq` in `Makefile` to change targets completely, I think you could avoid function like this:

```
server/static/files.go:
ifeq ($(STATIC_FILES),true)
  // do yarn stuff
else
  cp server/static/files.go.stub server/static/files.go
endif
```

Comment (simster7) [reply Makefile:134]: I'm thinking that I wouldn't want to use a build flag for this, and I think the current approach gives us more granularity in exactly where and when we create the mock file. Happy to use this pattern if you feel strongly about it though

Comment (alexec) [reply Makefile:134]: Please, unless you really cannot make it work.

Comment (simster7) [reply Makefile:134]: Will do

File: Makefile [hunk 7]
```diff
@@ -28,7 +28,12 @@ VERSION               := $(GIT_TAG)
 endif
 
 # should we build the static files?
+ifneq (,$(filter $@,codegen|lint|test))
+STATIC_FILES          := false
+else
```
Comment (simster7) [inline Makefile:33]: Do not build static files in the target is one of these

File: Makefile (hunk 6 above)
Comment (simster7) [reply Makefile]: Done

File: Makefile (hunk 7 above)
Comment (alexec) [reply Makefile]: nice - I did not know about that

File: Makefile [hunk 8]
```diff
@@ -28,7 +28,12 @@ VERSION               := $(GIT_TAG)
 endif
 
 # should we build the static files?
+ifneq (,$(filter $@,codegen|lint|test|docs))
+STATIC_FILES          := false
+else
```
Comment (simster7) [inline Makefile:33]: Do not build static files if target is one of these

File: Makefile [hunk 9]
```diff
@@ -155,24 +151,24 @@ images: cli-image executor-image controller-image
...
 $(GOPATH)/bin/staticfiles:
 	$(call go_install,bou.ke/staticfiles)
 
+ifeq ($(STATIC_FILES),true)
 server/static/files.go: $(GOPATH)/bin/staticfiles ui/dist/app/index.html
 	# Pack UI into a Go file.
 	$(GOPATH)/bin/staticfiles -o server/static/files.go ui/dist/app
+else
+server/static/files.go:
+	# Pack UI into a Go file.
```
Comment (alexec) [inline Makefile]: minor - comment incorrect

File: Makefile [hunk 10]
```diff
@@ -145,24 +151,23 @@ images: cli-image executor-image controller-image
...
 ui/dist/app/index.html: $(shell find ui/src -type f && find ui -maxdepth 1 -type f)
 	# `yarn install` is fast (~2s), so you can call it safely.
 	JOBS=max yarn --cwd ui install
 	# `yarn build` is slow, so we guard it with a up-to-date check.
 	JOBS=max yarn --cwd ui build
-else
-ui/dist/app/index.html:
-	@mkdir -p ui/dist/app
-	echo "Built without static files" > ui/dist/app/index.html
-endif
```
Comment (simster7) [inline Makefile:158]: No longer necessary as our stub file does not depend on `index.html`

File: Makefile [hunk 11]
```diff
@@ -368,12 +379,13 @@ lint: server/static/files.go $(GOPATH)/bin/golangci-lint
...
 	# Lint Go files
 	$(GOPATH)/bin/golangci-lint run --fix --verbose --concurrency 4 --timeout 5m
-	# Lint UI files
-ifeq ($(STATIC_FILES),true)
-	yarn --cwd ui lint
-endif
+
+	# Lint swagger files
+	swagger validate api/openapi-spec/swagger.json
+	go test ./api/openapi-spec
```
Comment (alexec) [inline Makefile:388]: can we remove this line - covered by `make test`?

File: server/static/files.go.stub [hunk 12]
```diff
@@ -0,0 +1,5 @@
+//DELETEME
```
Comment (alexec) [inline server/static/files.go.stub:1]: can delete this line?

**Review by alexec — APPROVED** [review]

File: Makefile (hunk 9 above)
Comment (simster7) [reply Makefile]: Will remove

File: Makefile (hunk 11 above)
Comment (simster7) [reply Makefile:388]: Will remove

File: server/static/files.go.stub (hunk 12 above)
Comment (simster7) [reply server/static/files.go.stub:1]: Will remove

**Pull request merged**